File: js/enemyManager.js
export class EnemyManager {
  constructor(game) {
    this.game = game;

    // Halved all base HP (50% of original)
    this.enemyBaseData = {
      drone: {
        baseHp: 15,
        gold: 5,
        baseSpeed: 80,
      },
      leaf_blower: {
        baseHp: 30,
        gold: 8,
        baseSpeed: 60,
      },
      trench_digger: {
        baseHp: 50,
        gold: 12,
        baseSpeed: 30,
      },
      trench_walker: {
        baseHp: 75,
        gold: 15,
        baseSpeed: 25,
      },
    };

    // Holds the loaded images, widths, heights, etc. from assetLoader.js
    this.loadedEnemyAssets = [];
  }

  setLoadedEnemyAssets(loadedEnemies) {
    this.loadedEnemyAssets = loadedEnemies;
  }

  spawnEnemy(type, hpMultiplier = 1) {
    const baseData = this.enemyBaseData[type] || this.enemyBaseData["drone"];
    const asset = this.loadedEnemyAssets.find(e => e.name === type)
      || this.loadedEnemyAssets[0];

    // 20% global HP reduction, then apply wave multiplier
    const finalHp = baseData.baseHp * 0.8 * hpMultiplier;

    // Random speed ~ ±20% around baseSpeed
    const speedFactor = 0.8 + Math.random() * 0.4;
    const finalSpeed = baseData.baseSpeed * speedFactor;

    const path = this.game.path;
    if (!path || path.length === 0) {
      console.warn("No path defined in Game; cannot spawn enemy properly.");
      return;
    }
    const firstWP = path[0];

    const enemy = {
      name: type,
      image: asset.image,
      width: asset.width,
      height: asset.height,
      x: firstWP.x,
      y: firstWP.y,
      hp: finalHp,
      baseHp: finalHp,
      speed: finalSpeed,
      gold: baseData.gold,
      waypointIndex: 1,
      dead: false,
    };

    this.game.enemies.push(enemy);
  }

  update(deltaSec) {
    // If gameOver, do not update enemies
    if (this.game.gameOver) return;

    // Move enemies, handle death
    this.game.enemies.forEach(e => {
      this.updateEnemy(e, deltaSec);
      if (e.hp <= 0) {
        this.game.gold += e.gold || 0;
        e.dead = true;
        // If this was the selected enemy, clear selection
        if (this.game.uiManager && this.game.uiManager.selectedEnemy === e) {
          this.game.uiManager.selectedEnemy = null;
          this.game.uiManager.hideEnemyStats();
        }
      }
    });

    // Remove dead or off-screen enemies
    this.game.enemies = this.game.enemies.filter(e => {
      if (e.dead) return false;
      if (e.x > this.game.width + e.width) {
        this.game.lives -= 1;
        if (this.game.lives <= 0 && !this.game.gameOver) {
          this.game.lives = 0; 
          this.game.endGame("Game Over");
        }
        return false;
      }
      return true;
    });
  }

  updateEnemy(enemy, deltaSec) {
    const path = this.game.path;
    const nextWP = path[enemy.waypointIndex];
    if (!nextWP) {
      // No next WP => just move off-screen
      enemy.x += enemy.speed * deltaSec;
      return;
    }

    // Move toward next waypoint
    const tx = nextWP.x;
    const ty = nextWP.y;
    const dx = tx - enemy.x;
    const dy = ty - enemy.y;
    const dist = Math.sqrt(dx * dx + dy * dy);
    const step = enemy.speed * deltaSec;

    if (dist <= step) {
      enemy.x = tx;
      enemy.y = ty;
      enemy.waypointIndex++;
    } else {
      enemy.x += (dx / dist) * step;
      enemy.y += (dy / dist) * step;
    }
  }

  drawEnemy(ctx, enemy) {
    this.drawImageSafely(
      ctx,
      enemy.image,
      enemy.x - enemy.width / 2,
      enemy.y - enemy.height / 2,
      enemy.width,
      enemy.height
    );

    // HP bar
    if (enemy.hp < enemy.baseHp) {
      const barW = enemy.width;
      const barH = 4;
      const pct  = Math.max(0, enemy.hp / enemy.baseHp);

      const barX = enemy.x - barW / 2;
      const barY = enemy.y - enemy.height / 2 - 6;

      ctx.fillStyle = "red";
      ctx.fillRect(barX, barY, barW, barH);

      ctx.fillStyle = "lime";
      ctx.fillRect(barX, barY, barW * pct, barH);
    }
  }

  drawImageSafely(ctx, image, x, y, w, h) {
    if (image.complete && image.naturalHeight !== 0) {
      ctx.drawImage(image, x, y, w, h);
    } else {
      console.warn("Image not loaded or invalid:", image);
    }
  }
}

File: python/enemy_manager.py
import pygame
import math
import random
import os

class EnemyManager:
    def __init__(self, game):
        self.game = game

        self.enemy_base_data = {
            "drone": {
                "baseHp": 30,
                "gold": 5,
                "baseSpeed": 80,
            },
            "leaf_blower": {
                "baseHp": 60,
                "gold": 8,
                "baseSpeed": 60,
            },
            "trench_digger": {
                "baseHp": 100,
                "gold": 12,
                "baseSpeed": 30,
            },
            "trench_walker": {
                "baseHp": 150,
                "gold": 15,
                "baseSpeed": 25,
            }
        }

        self.loaded_enemy_assets = {}
        for e_type in self.enemy_base_data.keys():
            image_path = os.path.join("assets", "enemies", f"{e_type}.png")
            if os.path.exists(image_path):
                img = pygame.image.load(image_path)
                max_dim = 30
                iw, ih = img.get_size()
                scale = max_dim / max(iw, ih)
                new_size = (int(iw * scale), int(ih * scale))
                scaled_img = pygame.transform.scale(img, new_size)
                self.loaded_enemy_assets[e_type] = scaled_img
            else:
                print(f"Warning: {image_path} not found.")
                surface = pygame.Surface((30,30))
                surface.fill((255,0,0))
                self.loaded_enemy_assets[e_type] = surface

    def update(self, delta_sec):
        to_remove = []
        for enemy in self.game.enemies:
            self.update_enemy(enemy, delta_sec)
            if enemy["hp"] <= 0:
                self.game.gold += enemy["gold"]
                to_remove.append(enemy)
            else:
                if enemy["waypointIndex"] >= len(self.game.path):
                    self.game.lives -= 1
                    if self.game.lives <= 0:
                        self.game.lives = 0
                        self.game.paused = True
                        print("Game Over")
                    to_remove.append(enemy)

        self.game.enemies = [e for e in self.game.enemies if e not in to_remove]

    def update_enemy(self, enemy, delta_sec):
        path = self.game.path
        if enemy["waypointIndex"] < len(path):
            next_wp = path[enemy["waypointIndex"]]
            ex, ey = enemy["x"], enemy["y"]
            tx, ty = next_wp
            dx, dy = tx - ex, ty - ey
            dist = math.hypot(dx, dy)
            step = enemy["speed"] * delta_sec

            if dist <= step:
                enemy["x"] = tx
                enemy["y"] = ty
                enemy["waypointIndex"] += 1
            else:
                enemy["x"] += (dx / dist) * step
                enemy["y"] += (dy / dist) * step

    def draw_enemy(self, screen, enemy):
        img = enemy["image"]
        rect = img.get_rect(center=(enemy["x"], enemy["y"]))
        screen.blit(img, rect)

        if enemy["hp"] < enemy["baseHp"]:
            bar_w = img.get_width()
            bar_h = 4
            pct = max(0, enemy["hp"] / enemy["baseHp"])
            bar_x = enemy["x"] - bar_w / 2
            bar_y = enemy["y"] - img.get_height() / 2 - 6
            pygame.draw.rect(screen, (255,0,0), (bar_x, bar_y, bar_w, bar_h))
            pygame.draw.rect(screen, (0,255,0), (bar_x, bar_y, bar_w * pct, bar_h))

    def spawn_enemy(self, e_type, hp_multiplier=1.0):
        base_data = self.enemy_base_data.get(e_type, self.enemy_base_data["drone"])
        asset = self.loaded_enemy_assets.get(e_type, self.loaded_enemy_assets["drone"])

        final_hp = base_data["baseHp"] * 0.8 * hp_multiplier
        speed_factor = 0.8 + random.random() * 0.4
        final_speed = base_data["baseSpeed"] * speed_factor

        if not self.game.path:
            print("No path defined, cannot spawn enemy!")
            return

        first_wp = self.game.path[0]
        enemy = {
            "name": e_type,
            "image": asset,
            "width": asset.get_width(),
            "height": asset.get_height(),
            "x": float(first_wp[0]),
            "y": float(first_wp[1]),
            "hp": final_hp,
            "baseHp": final_hp,
            "speed": final_speed,
            "gold": base_data["gold"],
            "waypointIndex": 1,
            "dead": False
        }
        self.game.enemies.append(enemy)
